### PR TITLE
HTTP API: Disallow GET requests

### PR DESCRIPTION
**Pull request by hsanjuan**
Better security by limiting the HTTP API to POST requests. (see commit descriptions).

### CODE REVIEW

**Comment by hsanjuan** [comment]
This might break tests, not sure, I will fix those on another day.

File: core/corehttp/commands.go [hunk 1]
```diff
@@ -140,15 +141,15 @@ func commandsOption(cctx oldcmds.Context, command *cmds.Command) ServeOption {
 }
 
 // CommandsOption constructs a ServerOption for hooking the commands into the
-// HTTP server.
+// HTTP server. These can be called only with POST, "OPTIONS" and "HEAD".
 func CommandsOption(cctx oldcmds.Context) ServeOption {
-	return commandsOption(cctx, corecommands.Root)
+	return commandsOption(cctx, corecommands.Root, []string{"POST", "OPTIONS", "HEAD"})
```
Comment (ribasushi) [inline core/corehttp/commands.go:146]: What is the rationale for allowing `OPTIONS` and `HEAD` on write opts?

Also do we not recognize `PUT`/`DELETE` anywhere?

Comment (hsanjuan) [reply core/corehttp/commands.go:146]: Ok, I was tired. I have removed HEAD (as this is akin to GET).

OPTIONS is necessary for CORS reasons, as it is how browsers figure out what's allowed and what not.

PUT/DELETE are not handled (from now on). In this API all request methods do the same thing.

**Comment by hsanjuan** [comment]
I have a test failure that seems unrelated: 

```
=== Failed
=== FAIL: core/coreapi/test TestIface/Dht/TestDhtFindPeer (1.29s)
panic: runtime error: index out of range [0] with length 0 [recovered]
	panic: runtime error: index out of range [0] with length 0

```

**Comment by hsanjuan** [comment]
@lidel can you help with the webui test failure (I have the feeling it will take you way shorter than me to figure out the place to fix it).

**Comment by lidel** [comment]
I will take a look as soon I get back home (eta 1h)

**Comment by hsanjuan** [comment]
Ok:

* Sharness is green

Pending:

* Reviews
* Need to merge and bubble https://github.com/ipfs/iptb-plugins/pull/37
* Need to fix webui test
* Final squash and fix commit messages.

File: core/corehttp/commands.go [hunk 2]
```diff
@@ -117,11 +117,17 @@ func patchCORSVars(c *cmdsHttp.ServerConfig, addr net.Addr) {
...
 }
 
-func commandsOption(cctx oldcmds.Context, command *cmds.Command) ServeOption {
+func commandsOption(cctx oldcmds.Context, command *cmds.Command, allowGet bool) ServeOption {
 	return func(n *core.IpfsNode, l net.Listener, mux *http.ServeMux) (*http.ServeMux, error) {
 
 		cfg := cmdsHttp.NewServerConfig()
-		cfg.SetAllowedMethods(http.MethodGet, http.MethodPost, http.MethodPut)
+		cfg.AllowGet = allowGet
+		corsAllowedMethods := []string{http.MethodPost, http.MethodPut}
```
Comment (Stebalien) [inline core/corehttp/commands.go:125]: I'm not currently allowing PUT in the commands lib. Will that be an issue?

Comment (lidel) [reply core/corehttp/commands.go:125]: AFAIK the only place where we use `PUT` is in "writable gateway",
which is on the Gateway port, but outside of `/api/v0` namespace.
Is there other use?

Comment (Stebalien) [reply core/corehttp/commands.go:125]: I don't think so. I'll remove this.

**Comment by lidel** [comment]
PSA: this will require IPFS Companion  v2.11.x or later (shipped a Beta for testing: [v2.11.0.904](https://github.com/ipfs-shipyard/ipfs-companion/releases/tag/v2.11.0.904))

**Comment by alexander255** [comment]
Of course I'll adapt to this in py-ipfs-http-client, but may I ask what the rationale is/was for blocking GET on idempotent & read-only API endpoints? Why can't things like `/api/v0/get?arg=<cid>` continue to be allowed GET access additionally to POST?

**Comment by Stebalien** [comment]
Basically, the HTTP API is not a REST API, it's a simple RPC API. In go-ipfs, at least, it's defined as a set of transport agnostic commands, then exposed through an HTTP API.

Note: You can continue to use GET requests against the read-only API on the gateway.

**Comment by alexander255** [comment]
Too bad :slightly_frowning_face:… I have some code deployed that uses the HTTP API as read-only API since its port is stable (5001) while the gateway port is unusable (8080 is way to generic). The decision to also enforce this for the read-only parts of the API parts that are also exposed via the gateway is final, I suppose? Does the read-only API exposed via the gateway support using POST or does only support GET instead?

**Comment by hsanjuan** [comment]
> Does the read-only API exposed via the gateway support using POST or does it only support GET instead?

I think both.

**Comment by Stebalien** [comment]
We have no intention of changing this as it simplifies reasoning about the system. Is it not possible for you to use POST?

**Comment by lidel** [comment]
@alexander255 @hsanjuan 👉 **`POST` restriction is only on the API port**

A subset of `/api/v0` exposed on the Gateway accepts `GET` just fine (0.5.0-rc2):

```console
$ curl -X GET http://127.0.0.1:5001/api/v0/dns/ipfs.io
405 - Method Not Allowed

$ curl -X GET http://127.0.0.1:8080/api/v0/dns/ipfs.io
{"Path":"/ipfs/bafybeibdwwmb725qy4yph56jg4bs6uimcst5hnvuraarowgdaytzmgsw24"}
```

**Comment by alexander255** [comment]
With the read-only API on gateway supporting POST, I can still pretend they are the same so no problem for me then. I just feels weird to use POST there, but if it makes stuff easier for you so be it. :+1:

**Pull request merged**